asterius/test/ghc-testsuite.hs

```diff
@@ -37,27 +64,169 @@ getTestCases = do
         readFileNullable (c -<.> "stdout") <*>
         readFileNullable (c -<.> "stderr")
 
-runTestCase :: TestCase -> IO (LBS.ByteString, LBS.ByteString)
+
+
+
+data TestOutcome = TestSuccess | TestFailure deriving(Eq, Show, Generic)
+
+instance ToJSON TestOutcome where
+
+
+data TestRecord = TestRecord
+  { trOutcome :: !TestOutcome
+  , trPath :: !FilePath -- ^ Path of the test case
+  , trErrorMessage :: !String -- ^ If the test failed, then the error message associated to the failure.
+  } deriving(Generic)
+
+instance ToJSON TestRecord where
+
+
+
+
+-- | Log of tests that have run
+newtype TestLog = TestLog [TestRecord] deriving(Semigroup, Monoid, Generic)
+
+instance ToJSON TestLog where
+
+-- | Append a value to the test log in safe way when we have multiple threads
+consTestLog :: TestRecord -> IORef TestLog -> IO ()
+consTestLog tr tlref = modifyIORef' tlref (\(TestLog tl) -> TestLog $ tr:tl)
```

Comment:
`atomicModifyIORef'` please. We might run the tests in parallel.

Reply:
whoops, my bad.